lib/types/ipblock.go

```diff
@@ -0,0 +1,173 @@
+package types
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"net"
+	"strings"
+)
+
+// ipBlock represents a continuous segment of IP addresses
+type ipBlock struct {
+	// TODO rename count as it is actually the accumulation of the counts up to and including this
+	// on in the pool
+	// maybe add another field, although technically we need count only to find out when to enter this
+	// block not for anything else
+	firstIP, count *big.Int
+	ipv6           bool
+}
+
+// IPPool represent a slice of IPBlocks
+type IPPool struct {
+	list  []ipBlock
+	count *big.Int
+}
+
+func getIPBlock(s string) (*ipBlock, error) {
+	switch {
+	case strings.Contains(s, "-"):
+		return ipBlockFromRange(s)
+	case strings.Contains(s, "/"):
+		return ipBlockFromCIDR(s)
+	default:
+		if net.ParseIP(s) == nil {
+			return nil, fmt.Errorf("%s is not a valid ip, range ip or CIDR", s)
+		}
+		return ipBlockFromRange(s + "-" + s)
+	}
+}
+
+func ipBlockFromRange(s string) (*ipBlock, error) {
+	ss := strings.SplitN(s, "-", 2)
+	ip0Str, ip1Str := strings.TrimSpace(ss[0]), strings.TrimSpace(ss[1])
+	ip0, ip1 := net.ParseIP(ip0Str), net.ParseIP(ip1Str)
+	if ip0 == nil || ip1 == nil {
+		fmt.Println("Wrong IP range format: ", s)
+		return nil, errors.New("wrong IP range format: " + s)
+	}
+	if (ip0.To4() == nil) != (ip1.To4() == nil) { // XOR
+		return nil, errors.New("mixed IP range format: " + s)
+	}
+	block := ipBlockFromTwoIPs(ip0, ip1)
+
+	if block.count.Sign() <= 0 {
+		return nil, errors.New("negative IP range: " + s)
+	}
+	return block, nil
+}
+
+func ipBlockFromTwoIPs(ip0, ip1 net.IP) *ipBlock {
+	var block ipBlock
+	block.firstIP = new(big.Int)
+	block.count = new(big.Int)
+	block.ipv6 = ip0.To4() == nil
+	if block.ipv6 {
+		block.firstIP.SetBytes(ip0.To16())
+		block.count.SetBytes(ip1.To16())
+	} else {
+		block.firstIP.SetBytes(ip0.To4())
+		block.count.SetBytes(ip1.To4())
+	}
+	block.count.Sub(block.count, block.firstIP)
```

Comment:
Please add a comment that ip0<ip1 is validated elsewhere, I missed the `negative ip range` check on the first reading